### ISSUE TITLE
Kinship coefficient calculation inconsistent with missing data

### ISSUE THREAD

**Issue by stephenturner** [Illumina/akt#33]
Hi @jaredo -- I realize you haven't committed to this repo in a few years, but just hoping you could point me in the right direction for fixing a problem I think I've discovered with the kinship calculation here. 

I have a VCF with two individuals, id1 and id2, (simulated parent-child). Each sample has a call rate of about 80%, so there are many spots where one sample is genotyped but the other isn't. When I run `akt kin` using the KING estimator with missing data present the kinship coefficient is underestimated compared to using `akt kin` when the missing SNPs are removed. In both examples akt correctly reports the actual number of genotypes compared (6238 out of the total 17,530).

The two-sample VCF with ~20% missing SNPs in each sample: [sim20mis_all.vcf.gz](https://github.com/Illumina/akt/files/6349703/sim20mis_all.vcf.gz)

When I run `akt kin` using the KING estimator, I underestimate the kinship coefficient at .19:

```
akt kin -M1 --force sim20mis_all.vcf.gz
```

```
id1        id2      0.00000 1.00000 0.00000 0.19066 6238
```

I can remove all the missing genotypes from this VCF (`bcftools view -g ^miss`): [sim20mis_nomiss.vcf.gz](https://github.com/Illumina/akt/files/6349715/sim20mis_nomiss.vcf.gz)

And re-run:

```
akt kin -M1 --force sim20mis_nomiss.vcf.gz
```

Which gives me the correct kinship coefficient, .24, for parent-offspring:

```
id1        id2      0.00000 1.00000 0.00000 0.24072 6238
```

I'm trying to go through the code in kin.cpp and elsewhere but I'm unsure where to start. Thanks for any help you might be able to provide!

**Comment by stephenturner** [comment]
One more note on this as well - when I run the actual KING software on the VCF with missing data, I get the correct kinship coefficient, .2407, instead of the erroneous kinship coefficient .19066.

**Comment by stephenturner** [comment]
I think the relevant code that needs to be modified is around here:

https://github.com/Illumina/akt/blob/3905c48e2d93b2c01edf6a81b9edce0a54bbf58d/kin.cpp#L197-L205

In looking at the code just above:
https://github.com/Illumina/akt/blob/3905c48e2d93b2c01edf6a81b9edce0a54bbf58d/kin.cpp#L183-L184

It seems that a modification to the king kinship coefficient loop should take into account whether the genotype is missing in one sample but not the other. Any help modifying this code would be greatly appreciated!

cc @rudyarthur

**Comment by stephenturner** [comment]
~~I think this may fix it: https://github.com/stephenturner/akt/commit/b0a7b65e41bfef2709015df4e9b1077b0e769717~~

Edit: that works for n=2 samples, but does not work with n>2 samples in the input VCF.

**Comment by jaredo** [comment]
Thanks for the detailed bug report Stephen. I will try to take a look soonish.

Sidenote: plink2 does have the KING estimator these days although I appreciate AKT's ability to using BCF + .csi still has utility.

**Comment by stephenturner** [comment]
Thanks @jaredo! Agreed, I'm a fan of AKT's API over Plink and/or KING. 

I tried poking around here, commented code out to leave the original implementation intact. This gives identical results to KING (or `akt kin -M1` if and only if there are no missing genotypes), **but only with n=2 samples in the vcf**. As soon as you have >=3, neither of these work any longer.

https://github.com/stephenturner/akt/blob/7c1bec82ead7261ba3541aaa163cba0d3a9facc4/kin.cpp#L195-L223

**Comment by stephenturner** [comment]
Also, if it helps @jaredo I have a large multisample VCF, 5 generations, known pedigree, simulated VCF, with 20% missing data. KING implementation if KING algorithm and the Plink2 implementation of the KING algorithm provide the same results, where `akt kin -M1` underestimates. The VCF is ~68M. I can put in in a dropbox or something if that'd help.

**Comment by stephenturner** [comment]
For what it's worth, here's the comparison of the kinship coefficient of the `akt -M1` implementation of king against the actual KING software implementation (left) and plink2 implementation of the king algorithm against KING (right), in the presence of randomly missing data, on this 5-generation pedigree.

![image](https://user-images.githubusercontent.com/460076/115788220-55342c00-a391-11eb-994f-d9781abae19e.png)

**Comment by jaredo** [comment]
Hi @stephenturner. I think this is fixed. Could you test on my personal repo?

https://github.com/jaredo/akt

Thanks for reporting this, the bug was as you expected, we were not correctly handling missingness when counting the number of hets in an individual. I guess this has gone unnoticed since I normally use data with very low amounts of missing genotypes.

If that looks okay to you, I will get this merged into Illumina repo. cc @olest @rudyarthur 

PS see https://github.com/jaredo/akt/commit/9e21cef6e9bb67795651579169780130070331b0 for the specific fix

**Comment by stephenturner** [comment]
Looks like this is fixed once you merge! (k.akt1=original implementation, k.akt2=your fix with the mask)

![image](https://user-images.githubusercontent.com/460076/116061065-6fe7f880-a650-11eb-8da2-5562b97c2537.png)

**Comment by jaredo** [comment]
@stephenturner you should be good to go

**Comment by stephenturner** [comment]
Fixed in 9e21cef6e9bb67795651579169780130070331b0 via 4c91108636fdc9bdc40a4ef23dc2036d48514dd8 thanks @jaredo @olest
